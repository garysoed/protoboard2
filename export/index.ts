export {Lens, $lens} from '../src/core/lens';
export {LensDisplay} from '../src/core/lens-display';

// action
export {pickAction} from '../src/action/pick-action';

// coordinate
export {Cartesian} from '../src/coordinate/cartesian';
export {Coordinate} from '../src/coordinate/coordinate';
export {Indexed} from '../src/coordinate/indexed';

// region
export {Deck, $deck, DeckSpec} from '../src/region/deck';
export {Slot, $slot, SlotSpec, slotSpec} from '../src/region/slot';

// core
export {$createSpecEntries, $createSpecMap} from '../src/objects/object-create-spec';
export {ContentSpec} from '../src/payload/is-container';
export {ObjectSpec} from '../src/types/object-spec';
export {$render, Render} from '../src/objects/render';
export {$active, Active, activeSpec} from '../src/core/active';
export {$$activeSpec} from '../src/objects/active-spec';

// piece
export {$d1, D1, D1Spec, d1Spec} from '../src/piece/d1';
export {$d2, D2, D2Spec} from '../src/piece/d2';
export {$d6, D6, D6Spec} from '../src/piece/d6';

// tiling
export {rectOrthogonal, RectOrthogonal, Direction as RectOrthogonalDirection, distance as rectOrthogonalDistance} from '../src/tiling/rect-orthogonal';
export {vhex, VHex, Direction as VHexDirection, distance as vhexDistance} from '../src/tiling/vhex';
export {Tile} from '../src/tiling/types';
