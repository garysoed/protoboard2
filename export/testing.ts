export {LensHarness} from '../src/face/testing/lens-harness';
export {createRenderSpec, TEST_FACE} from '../src/testing/test-face';