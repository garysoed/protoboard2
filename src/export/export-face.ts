import {assertByType} from 'gs-tools/export/rxjs';
import {instanceofType} from 'gs-types';
import {RenderSpec} from 'persona';
import {render, RenderContext} from 'persona/export/internal';
import {from, OperatorFunction, pipe} from 'rxjs';
import {map, switchMap} from 'rxjs/operators';


interface Dimension {
  readonly height: number;
  readonly width: number;
}

export function exportFace(dimension: Dimension, context: RenderContext): OperatorFunction<RenderSpec, string> {
  const canvas = document.createElement('canvas');
  canvas.width = dimension.width;
  canvas.height = dimension.height;

  const context2d = canvas.getContext('2d');
  if (!context2d) {
    throw new Error('2d context could not be created');
  }

  return pipe(
      switchMap(renderSpec => render(renderSpec, context)),
      assertByType(instanceofType(SVGElement)),
      switchMap(svg => {
        const svgData = `data:image/svg+xml;base64,${btoa(svg.outerHTML)}`;
        const image = new Image(dimension.width, dimension.height);
        image.src = svgData;

        return from(image.decode()).pipe(map(() => image));
      }),
      map(image => {
        context2d.drawImage(image, 0, 0);
        return canvas.toDataURL('image/png');
      }),
  );
}