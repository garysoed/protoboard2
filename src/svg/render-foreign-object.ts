import {ocase, query, renderString, FOREIGN_OBJECT, DIV, RenderSpec, ParseType} from 'persona';
import {ElementNamespace} from 'persona/export/internal';
import {Observable, OperatorFunction, of} from 'rxjs';
import {map} from 'rxjs/operators';

export interface RenderArgs {
  readonly rect$: Observable<SVGRectElement>;
  readonly renderContent: OperatorFunction<unknown, RenderSpec>;
}

export function renderForeignObject(args: RenderArgs): RenderSpec {
  return renderString({
    raw: of(`<foreignObject><div xmlns="${ElementNamespace.HTML}"></div></foreignObject>`),
    spec: {
      content: query('div', DIV, {
        content: ocase<{}>(),
      }),
      foreignObject: query(null, FOREIGN_OBJECT, {}),
    },
    runs: $foreign => [
      of({}).pipe($foreign.content.content(args.renderContent)),
      args.rect$.pipe(map(target => target.x.baseVal.value), $foreign.foreignObject.x()),
      args.rect$.pipe(map(target => target.y.baseVal.value), $foreign.foreignObject.y()),
      args.rect$.pipe(map(target => target.width.baseVal.value), $foreign.foreignObject.width()),
      args.rect$.pipe(map(target => target.height.baseVal.value), $foreign.foreignObject.height()),
    ],
    parseType: ParseType.SVG,
  });
}

