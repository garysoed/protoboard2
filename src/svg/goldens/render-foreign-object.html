<pbt-test><gst-fake-shadow>
  <svg>
    <rect fill="grey" x="20" y="40" width="60" height="100"></rect>
    <g><foreignObject x="20" y="40" width="60" height="100"><div xmlns="http://www.w3.org/1999/xhtml"><div>
        <style>
          .red {
            background-color: red;
            position: absolute;
            left: 10px;
            top: 20px;
          }

          .green {
            background-color: green;
            position: absolute;
            left: 30px;
            top: 40px;
          }

          .box {
            height: 40px;
            width: 40px;
          }
        </style>

        <div class="red box"></div>
        <div class="green box"></div>
      </div></div></foreignObject></g>
  </svg>
  </gst-fake-shadow></pbt-test>